include/aspect/simulator/assemblers/interface.h

```diff
@@ -0,0 +1,718 @@
+                internal::Assembly::CopyData::CopyDataBase<dim> &data) const =0;
+
+        /**
+         * This function gets called if a MaterialModelOutputs is created
+         * and allows the assembler to attach AdditionalOutputs. The
+         * function might be called more than once for a
+         * MaterialModelOutput, so it is recommended to check if
+         * get_additional_output() returns an instance before adding a new
+         * one to the additional_outputs vector. By default this function does
+         * not create additional outputs.
+         *
+         * Material models, through functions derived from
+         * MaterialModel::Interface::evaluate(), put their computed material
+         * parameters into a structure of type MaterialModel::MaterialModelOutputs.
+         * By default, material models will compute those parameters that
+         * correspond to the member variables of that structure. However,
+         * there are situations where parts of the simulator need additional
+         * pieces of information; a typical example would be the use of a
+         * Newton scheme that also requires the computation of <i>derivatives</i>
+         * of material parameters with respect to pressure, temperature, and
+         * possibly other variables.
+         *
+         * The computation of such additional information is controlled by
+         * the presence of a collection of pointers in
+         * MaterialModel::MaterialModelOutputs that point to additional
+         * objects. Whether or not one needs these additional objects depends
+         * on what assemblers are selected, or what postprocessing one
+         * wants to compute. For the purpose of assembly, the current
+         * function creates the additional objects (such as the one that stores
+         * derivatives) and adds pointers to them to the collection, based on
+         * what this assembler class requires. This function is always called
+         * before the material model is evaluated and execute() is called.
+         * This ensures the additional material model output is available when
+         * execute() is called.
+         */
+        virtual void create_additional_material_model_outputs(MaterialModel::MaterialModelOutputs<dim> &) const {}
+
+        /**
+         * A required function for objects that implement the assembly of terms
+         * in an equation that requires the computation of residuals
+         * (in particular the advection equation in ASPECT).
+         * Just like the assemblers itself, the residual
+         * that we use to compute the necessary entropy viscosity depend on the
+         * equation (i.e. which terms are actually included in the
+         * equation). Thus different objects compute different residuals (i.e.
+         * the residual for a melt advection equation looks different from the
+         * residual for a passive compositional field).
```

Comment:
add "For assemblers for the Stokes system, an implementation of this function is not necessary"